providers/linux/vmstat.go

```diff
@@ -0,0 +1,63 @@
+// Licensed to Elasticsearch B.V. under one or more contributor
+// license agreements. See the NOTICE file distributed with
+// this work for additional information regarding copyright
+// ownership. Elasticsearch B.V. licenses this file to you under
+// the Apache License, Version 2.0 (the "License"); you may
+// not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+package linux
+
+import (
+	"reflect"
+
+	"github.com/pkg/errors"
+
+	"github.com/elastic/go-sysinfo/types"
+)
+
+// parseVMStat parses the contents of /proc/vmstat
+func parseVMStat(content []byte) (*types.VMStatInfo, error) {
+	vmStat := &types.VMStatInfo{}
+	refVal := reflect.ValueOf(vmStat).Elem()
+
+	err := parseKeyValue(content, " ", func(key, value []byte) error {
+		// turn our []byte value into an int
+		val, err := parseBytesOrNumber(value)
+		if err != nil {
+			return errors.Wrapf(err, "failed to parse %v value of %v", string(key), string(value))
+		}
+
+		// Search The struct object to see if we have a field with a tag that matches the raw key coming off the file input
+		// This is the best way I've found to "search" for a a struct field based on a struct tag value.
+		// In this case, the /proc/vmstat keys are struct tags.
+		fieldToSet := refVal.FieldByNameFunc(func(name string) bool {
```

Comment:
Perhaps it would make sense to iterate over the fields and read their tags just once to build up a map of tag to Value. This looks likes it's O(n^2) now.

Reply:
Ugh, yah, I think you're right. This is gonna end up getting called by metricbeat on every Fetch, so I'll see what I can do.